### ISSUE TITLE
"Solving environment: failed" after conda update conda

### ISSUE THREAD

**Issue by davidechicco** [bioconda/bioconda-recipes#7950]
Hi
I was trying to solve a p[roblem I've been having with bioconda](https://github.com/bioconda/bioconda-recipes/issues/7920). I tried to remove Minoconda2 and reinstall it. Everything  seemed to work alright, but now if I run `conda update conda`, or `conda update --all`, or `conda install -c bioconda behst`, I get this error:

```
$ conda update conda
Solving environment: failed

NotWritableError: The current user does not have write permissions to a required path.
  path: /path/to/custom/dir/pkgs/urls.txt
  uid: 1000
  gid: 1000

If you feel that permissions on this path are set incorrectly, you can manually
change them by executing

  $ sudo chown 1000:1000 /path/to/custom/dir/pkgs/urls.txt

In general, it's not advisable to use 'sudo conda'.
```
Does anyone know how to solve this problem? Thanks!

**Comment by rvalieris** [comment]
> NotWritableError: The current user does not have write permissions to a required path.

conda is pointing out that there are files inside the conda installation that you don't have permission to write.
this shouldn't happen, only your user should be interacting with your conda prefix.

you can use chown to change the permissions back, as the message says.

**Comment by davidechicco** [comment]
Thanks

**Comment by davidechicco** [comment]
Hi
I have this problem again...  what are the files at /path/to/custom/dir/pkgs/urls.txt to which the message refers?

Thanks

-- Davide

**Comment by PertuyF** [comment]
I guess is stores the exact urls for each package you install.
I used to have the same error in a multi-user install, if one user installs a package this file get updated. During this process the permissions can be changed and prevent modification by another user.

Could it be you case too?

**Comment by mbargull** [comment]
There have been some changes in regards to how permissions are handled for the package cache, see https://github.com/conda/conda/issues/7267#issuecomment-420571523. I didn't check out what changes have been made, so can't comment further on it. But if you want, you can try and install the latest prerelease version, `conda-canary::conda=4.6.0b0`, and report back in https://github.com/conda/conda/issues/7267.

**Comment by davidechicco** [comment]
@PertuyF Thanks for your help, but I'm the only user on my laptop. Could it me a mess between my username (davide) and the root username? 
@mbargull Thanks. How can I install the latest pre-release version?

**Comment by PertuyF** [comment]
@davidechicco to explore if it is a file owner problem, try `ls -hla $CONDA_PREFIX/pkgs/urls.txt` and look at the owner and owner group. But if you're the only one using this laptop with only your account, it is less likely.

To install the beta version as suggested by @mbargull , just use `conda install conda-canary::conda=4.6.0b0`. This syntax will try to install `conda` from the `conda-canary` channel, where the beta versions are hosted. In this case it is similar to `conda install -c conda-canary conda=4.6.0b0`.

**Comment by mbargull** [comment]
Yes, `conda install conda-canary::conda=4.6.0b0` or just `conda install conda-canary::conda` is correct.
(It is similar but not equivalent to `-c conda-canary`: with `conda-canary::conda` you are telling `conda` to install that package from `conda-canary` and at the same time consider other packages from `conda-canary` only with lowest priority. With `-c conda-canary` you are telling `conda` to give all packages from `conda-canary` highest priority. There are some older versions of some very few packages on `conda-canary`, though, that you likely don't want. Hence it is usually better to install with `conda-canary::conda` instead. I know, I know, the UI for this isn't terrible good and obvious here :wink:.)

**Comment by davidechicco** [comment]
Hi @PertuyF and @mbargull 
I tried `conda install conda-canary::conda=4.6.0b0` and `conda install conda-canary::conda` but I always get the same mistake :-(

**Comment by mbargull** [comment]
Ah right, to be able to update you have to workaround the issue, of course, sorry.
Do you have the `pkgs_dirs` parameter set in your `conda` config? If not, you can simply try
```
CONDA_PKGS_DIRS=/some/temporary/path conda install conda-canary::conda
```
for that update.

**Comment by davidechicco** [comment]
Thanks for your help @mbargull. I created the `./temp18`folder and tried to run `CONDA_PKGS_DIRS=temp18 conda install conda-canary::cond` but I got the same error again... :-(

**Comment by PertuyF** [comment]
You can try a `chmod 666 $CONDA_PREFIX/pkgs/urls.txt`, which should solve the problem for the `conda` update. Then change the permissions back to their initial state eventually.

**Comment by davidechicco** [comment]
Thanks @PertuyF, but my $CONDA_PREFIX is empty

**Comment by PertuyF** [comment]
Currently you have activate a conda environment before using this variable, so try `conda activate base`.
Otherwise, replace it with the absolute path to your conda install dir, `pkgs/urls.txt` is usually sitting right there.

It is the file mentioned in your install error that should be modified.

**Comment by mbargull** [comment]
`CONDA_PREFIX` points to the currently activated environment. The package cache is not generally under `${CONDA_PREFIX}`. If `pkgs_dirs` is not configured to point elsewhere, `"$(conda info --base)"/pkgs` and `~/.conda/pkgs` are the standard paths for it.

@davidechicco, there is insufficient information for me to see where the problem comes from. I'd at least need output from
```sh
conda --version
CONDA_PKGS_DIRS=/tmp conda config --show pkgs_dirs
conda config --show pkgs_dirs --json | python -c 'from sys import stdin; from json import loads; print("\n".join(loads(stdin.read())["pkgs_dirs"]))' | xargs -I'$' ls -ld '$' '$/urls.txt'
```

**Comment by davidechicco** [comment]
Thanks @PertuyF and @mbargull for your help. Here are the output messages obtained on my laptop (Dell Latitude 3450 running a Linux CentOS 7 operating system) right now.

Command: `conda --version`
Output: `conda 4.5.11`

Command: `CONDA_PKGS_DIRS=/tmp conda config --show pkgs_dirs`
Output:
`pkgs_dirs:`
`  - /tmp`
`  - /tmp/tmp`
`  - /path/to/custom/dir/pkgs`

Command: `conda config --show pkgs_dirs --json | python -c 'from sys import stdin; from json import loads; print("\n".join(loads(stdin.read())["pkgs_dirs"]))' | xargs -I'$' ls -ld '$' '$/urls.txt'`
Output:
`drwxrwxr-x. 12 davide davide 4096 Sep 25 18:35 /tmp/tmp`
`-rw-rw-r--.  1 davide davide  680 Sep 25 18:35 /tmp/tmp/urls.txt`
`ls: cannot access /path/to/custom/dir/pkgs: No such file or directory`
`ls: cannot access /path/to/custom/dir/pkgs/urls.txt: No such file or directory`

I hope you can help me, thanks again!

**Comment by PertuyF** [comment]
Looks like `/path/to/custom/dir/pkgs` may not exist at all. Otherwise try `ls -hla /path/to/custom/dir/pkgs/` to see the exact permissions.

**Comment by davidechicco** [comment]
@PertuyF  Yes, `/path/to/custom/dir/pkgs` does not exist. What can i do?

Thanks

**Comment by mbargull** [comment]
`conda config --show-sources` will tell you which config file injected that path. You'll have to remove it from there.

**Comment by Aisuko** [comment]
The same issue with you guys, First time i was used `sudo sh .Miniconda....sh`, so, this situation was happen to me, and the miniconda3 folder permission is root. Then, i had to uninstall miniconda3, and reinstall miniconda3 with command `sh .Miniconda..sh`, and it work for me, the miniconda3 folder permission is local user. So, I think change folder permission maybe useful.

**Comment by KseniaErshova** [comment]
This is what helped me:
`sudo chown -R user anaconda3`

**Comment by Jiceh3** [comment]
Oui moi aussi ça m'a aidé. C'est le code qui donne la permission à l'utilisateur user sur le répertoire anaconda3
il faut remplacer user par le nom d'utilisateur.

> sudo chown -R user anaconda3

**Comment by MuhammadBilal96** [comment]
> Oui moi aussi ça m'a aidé. C'est le code qui donne la permission à l'utilisateur user sur le répertoire anaconda3
> il faut remplacer user par le nom d'utilisateur.
> 
> > sudo chown -R user anaconda3

u r a life saver

**Comment by KayAiIn** [comment]
> This is what helped me:
> `sudo chown -R user anaconda3`

this really worked for me, thanks

**Comment by rjolicoe** [comment]
This is what helped me:
sudo chown -R user anaconda3

This really worked for me, thanks!

**Comment by pakaraima** [comment]
> This is what helped me:
> `sudo chown -R user anaconda3`

This solved my problem. Thanks.

**Comment by gedeon1310** [comment]
Elegant solution. Worked for me, much appreciated !!

**Comment by anshumankmr** [comment]
Literally none of these solutions worked for me.

**Comment by PertuyF** [comment]
Then your issue may be different, please consider creating a new one.

**Comment by anshumankmr** [comment]
The traceback is pretty much the same, but I will make one.

**Comment by just1ndca** [comment]
> This is what helped me:
> `sudo chown -R user anaconda3`
It works 4 me, ur god

**Comment by jckuri** [comment]
> This is what helped me:
> `sudo chown -R user anaconda3`

Thank you!
This worked for me:
`sudo chown -R MY_USER ~/.conda`

**Comment by greeneley** [comment]
Nice. This exact command worked for me:
```
sudo chrown -R $USER ~/.conda
```

**Comment by swipswaps** [comment]
//this?
sudo chown -R $USER ~/.conda

**Comment by WebReflection** [comment]
to whome it might concern, I've used `sudo chown -R root:users /opt/miniconda3` in ArchLinux after miniconda3 AUR package and everything seemed to work just fine after setting up an env.
